legend-engine-server/src/test/java/org/finos/legend/engine/server/test/shared/TestServer.java

```diff
@@ -0,0 +1,88 @@
+// Copyright 2020 Goldman Sachs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.finos.legend.engine.server.test.shared;
+
+import io.dropwizard.setup.Environment;
+import org.finos.legend.engine.plan.execution.stores.relational.connection.tests.api.ConnectionTestManager;
+import org.finos.legend.engine.plan.execution.stores.relational.connection.tests.api.DynamicTestConnection;
+import org.finos.legend.engine.protocol.pure.v1.model.packageableElement.store.relational.connection.DatabaseType;
+import org.finos.legend.engine.protocol.pure.v1.model.packageableElement.store.relational.connection.RelationalDatabaseConnection;
+import org.finos.legend.engine.server.Server;
+import org.finos.legend.engine.server.ServerConfiguration;
+import org.finos.legend.engine.shared.core.url.EngineUrlStreamHandlerFactory;
+
+import java.util.*;
+
+public class TestServer extends Server
+{
+    private final List<DynamicTestConnection> dynamicTestConnections = new ArrayList<>();
+
+    public static void main(String[] args) throws Exception
+    {
+        EngineUrlStreamHandlerFactory.initialize();
+        new TestServer().run(args);
+    }
+
+    @Override
+    public void run(ServerConfiguration serverConfiguration, Environment environment)
+    {
+        super.run(serverConfiguration, environment);
+        environment.jersey().register(new TestConnectionProviderApi(getTestConnections(serverConfiguration)));
+    }
+
+    public Map<DatabaseType, RelationalDatabaseConnection> getTestConnections (ServerConfiguration serverConfiguration)
+    {
+        Map<DatabaseType, RelationalDatabaseConnection> testConnections = new HashMap<>();
+
+        // register static test connections
+        for (DatabaseType dbType : serverConfiguration.staticTestConnections.keySet())
+        {
+            if (serverConfiguration.testConnectionsToEnable.contains(dbType))
+            {
+                testConnections.putIfAbsent(dbType, serverConfiguration.staticTestConnections.get(dbType));
+            }
+        }
+
+        //Spin up dynamic database instances for testing
+        for (ConnectionTestManager connectionTestManager : ServiceLoader.load(ConnectionTestManager.class))
```

Comment:
The connection test manager resolution is non-deterministic.

Say I want to run the test framework and I don't have access to run SqlServer via Docker. 
I can implement my own manager but this loop might pick up the Docker test manager from Legend and fail. 

We have to make this deterministic.

Reply:
It is deterministic if you include the following assumptions.
1. Static test connections provided in server config take precedence over dynamic test connections
2. A single ConnectionTestManager exists per dbType, which can optionally provide a dynamically created test db's connection. The main purpose of ConnectionTestManager is to provide info about db specific tests for various db types along with the ability to launch an H2 type equivalent for them.

Reply:
connection implementation to use needs to be defined exclusively in config now. TestServer only picks the specified implementation of dynamic connection in its path